determinism.md

```diff
@@ -0,0 +1,171 @@
+GetDeterministicKey
+===
+
+**This is meant to be the text of the issue I will eventually file for the API**
+
+## Background and Motivation
+The `Compilation` type is fully deterministic meaning that given the same inputs
+(`SyntaxTree`, `CompilationOptions`, etc ...) it will produce the same output. By 
+that it means it will produce byte for byte equivalent binaries and the same 
+diagnostics. This is extremely valuable in infrastructure because it allows 
+for caching and opens the door for efficient distributed processing.
+
+At the moment this is hard to leverage because consumers can't determine if 
+two `Compilation` instances are equivalent for the purposes of determinism. 
+Customers have to resort to hand written comparisons which requires a fairly 
+intimate knowledge of the compiler (for example knowing what does and does not
+impact determinism). Such solutions are not version tolerant; every time the 
+compiler adds a new property that impacts determinism the solution must be 
+updated. Even when proper equality checks are in place this does not help 
+distributed computing where equality must be decided across different processes.
+
+The motivation here is to provide an API that returns a string based key for a
+given `Compilation` such that for two equivalent `Compilation` instances the
+key will also be equivalent. This will allow customers who want to leverage 
+deterministic caching and communication to have a simple, and transmittable,
+way to full describe the `Compilation` they are processing.
+
+## Proposed API
+
+```diff
+namespace Microsoft.CodeAnalysis
+{
+    public class Compilation
+    {
++       public string GetDeterministicKey(
++           ImmutableArray<AdditionalText> additionalTexts = default,
++           ImmutableArray<DiagnosticAnalyzer> analyzers = default,
++           ImmutableArray<ISourceGenerator> generators = default,
++           EmitOptions? emitOptions = null,
++           DeterministicKeyOptions options = DeterministicKeyOptions.Default)
```

Comment:
can we update to have hte static entrypoint?

Reply:
I've been trying to think on what the best `static` entry point would be. The implementation of the key can't use `CompilationOptions` for example it needs to use values off of `CSharpCompilationOptions`. One of the advantages of having `Compilation.GetDeterministicKey` is the implementation was in the CSharp or Basic versions which knew what type to cast `CompilationOptions` too. 

Once there is a `static` entry point we lose that virtual dispatch a little bit. I can still add it back in the implementation it is just a bit less pretty. I will take a stab at the design and see what everyone thinks but open to early suggestions as well.

Reply:
Okay updated with a `static` entry point here sketched out.